File: ui-tests/PageObjects/Panes/CallTrace/CallTraceEntry.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Playwright;
using UiTests.PageObjects.Components;
using UiTests.Utils;

namespace UiTests.PageObjects.Panes.CallTrace;

/// <summary>
/// Represents a single call entry rendered within the call trace pane.
/// </summary>
public class CallTraceEntry
{
    private static readonly string[] ExpandChildrenOptions = { "Expand Call Children", "Expand Full Callstack" };
    private static readonly string[] CollapseChildrenOptions = { "Collapse Call Children", "Expand Full Callstack" };

    private readonly CallTracePane _pane;
    private readonly ILocator _root;
    private readonly ContextMenu _contextMenu;

    public CallTraceEntry(CallTracePane pane, ILocator root, ContextMenu contextMenu)
    {
        _pane = pane ?? throw new ArgumentNullException(nameof(pane));
        _root = root ?? throw new ArgumentNullException(nameof(root));
        _contextMenu = contextMenu ?? throw new ArgumentNullException(nameof(contextMenu));
    }

    /// <summary>
    /// Root locator for the entry (<c>.calltrace-call-line</c>).</summary>
    public ILocator Root => _root;

    private ILocator OffsetLocator() => _root.Locator("> span").First;
    private ILocator ChildBoxLocator() => _root.Locator(".call-child-box");
    private ILocator CallTextLocator() => ChildBoxLocator().Locator(".call-text");
    private ILocator ToggleLocator() => ChildBoxLocator().Locator(".toggle-call");
    private ILocator ReturnLocator() => ChildBoxLocator().Locator(".return-text");
    private ILocator ArgumentContainer() => ChildBoxLocator().Locator(".call-args");

    /// <summary>
    /// Extracts the raw call text (e.g. "iterate_asteroids #1").
    /// </summary>
    public async Task<string> CallTextAsync()
    {
        var text = await CallTextLocator().InnerTextAsync();
        return text?.Trim() ?? string.Empty;
    }

    /// <summary>
    /// Function name portion extracted from the call text.
    /// </summary>
    public async Task<string> FunctionNameAsync()
    {
        var callText = await CallTextAsync();
        var hashIndex = callText.IndexOf('#');
        return hashIndex >= 0 ? callText[..hashIndex].Trim() : callText;
    }

    /// <summary>
    /// Call identifier (the suffix after '#', if present).
    /// </summary>
    public async Task<string> CallIdentifierAsync()
    {
        var callText = await CallTextAsync();
        var hashIndex = callText.IndexOf('#');
        return hashIndex >= 0 ? callText[(hashIndex + 1)..].Trim() : string.Empty;
    }

    /// <summary>
    /// Depth inferred from the leading spacer element.
    /// </summary>
    public async Task<int> DepthAsync()
    {
        var style = await OffsetLocator().GetAttributeAsync("style") ?? string.Empty;
        var minWidthPrefix = "min-width:";
        var index = style.IndexOf(minWidthPrefix, StringComparison.OrdinalIgnoreCase);
        if (index < 0)
        {
            return 0;
        }

        var start = index + minWidthPrefix.Length;
        var end = style.IndexOf('p', start);
        if (end < 0)
        {
            end = style.Length;
        }

        var valueString = style[start..end].Trim().TrimEnd('x');
        if (!double.TryParse(valueString, NumberStyles.Any, CultureInfo.InvariantCulture, out var pixels))
        {
            return 0;
        }

        return (int)Math.Round(pixels / 8.0, MidpointRounding.AwayFromZero);
    }

    private async Task<bool> HasClassAsync(string className)
    {
        var classAttr = await _root.GetAttributeAsync("class") ?? string.Empty;
        return classAttr.Split(' ', StringSplitOptions.RemoveEmptyEntries)
            .Contains(className, StringComparer.OrdinalIgnoreCase);
    }

    /// <summary>
    /// Indicates whether this entry is the currently selected call.
    /// </summary>
    public Task<bool> IsSelectedAsync() => HasClassAsync("event-selected");

    /// <summary>
    /// Determines whether the entry exposes a toggle button.
    /// </summary>
    public async Task<bool> HasToggleAsync()
    {
        return await ToggleLocator().Locator(".collapse-call-img, .dot-call-img").CountAsync() > 0;
    }

    /// <summary>
    /// Checks if child calls are currently expanded.
    /// </summary>
    public async Task<bool> HasExpandedChildrenAsync()
    {
        return await ToggleLocator().Locator(".collapse-call-img").CountAsync() > 0;
    }

    /// <summary>
    /// Returns the call context-menu options expected for the current state.
    /// </summary>
    public async Task<IReadOnlyList<string>> ExpectedContextMenuAsync()
    {
        return await HasExpandedChildrenAsync() ? CollapseChildrenOptions : ExpandChildrenOptions;
    }

    /// <summary>
    /// Opens the call context menu and retrieves the visible entries.
    /// </summary>
    public async Task<IReadOnlyList<string>> ContextMenuEntriesAsync()
    {
        await ChildBoxLocator().ClickAsync(new() { Button = MouseButton.Right });
        await _contextMenu.WaitForVisibleAsync();
        var entries = await _contextMenu.GetEntriesAsync();
        await _contextMenu.DismissAsync();
        return entries.Select(e => e.Text).ToList();
    }

    /// <summary>
    /// Expands child calls when possible.
    /// </summary>
    public async Task ExpandChildrenAsync()
    {
        if (!await HasToggleAsync())
        {
            return;
        }

        if (!await HasExpandedChildrenAsync())
        {
            await ToggleLocator().ClickAsync();
            await RetryHelpers.RetryAsync(HasExpandedChildrenAsync);
        }
    }

    /// <summary>
    /// Collapses the children when expanded.
    /// </summary>
    public async Task CollapseChildrenAsync()
    {
        if (!await HasToggleAsync())
        {
            return;
        }

        if (await HasExpandedChildrenAsync())
        {
            await ToggleLocator().ClickAsync();
            await RetryHelpers.RetryAsync(async () => !await HasExpandedChildrenAsync());
        }
    }

    /// <summary>
    /// Clicks the call entry to trigger jump navigation.
    /// Tries multiple hit targets (call text, child container, root) to accommodate
    /// renderer differences between Electron and Web.
    /// </summary>
    public async Task ActivateAsync()
    {
        var functionName = await FunctionNameAsync();
        DebugLogger.Log($"CallTraceEntry[{functionName}]: Begin ActivateAsync");

        var targets = new (string Label, ILocator Locator)[]
        {
            ("call-text", CallTextLocator()),
            ("child-box", ChildBoxLocator()),
            ("root", _root)
        };

        var clickAttempts = new (string Label, LocatorClickOptions Options)[]
        {
            ("single-click", new LocatorClickOptions { ClickCount = 1 }),
            ("double-click", new LocatorClickOptions { ClickCount = 2, Delay = 50 }),
            ("forced-click", new LocatorClickOptions { ClickCount = 1, Force = true })
        };

        foreach (var (targetLabel, target) in targets)
        {
            try
            {
                DebugLogger.Log($"CallTraceEntry[{functionName}]: scrolling into view {targetLabel}");
                await target.ScrollIntoViewIfNeededAsync();
            }
            catch (PlaywrightException ex)
            {
                DebugLogger.Log($"CallTraceEntry[{functionName}]: scroll failed for {targetLabel}: {ex.Message}");
            }

            foreach (var (clickLabel, options) in clickAttempts)
            {
                try
                {
                    DebugLogger.Log($"CallTraceEntry[{functionName}]: clicking {targetLabel} with {clickLabel}");
                    await target.ClickAsync(options);
                    try
                    {
                        await RetryHelpers.RetryAsync(async () => await IsSelectedAsync(), maxAttempts: 5, delayMs: 50);
                        DebugLogger.Log($"CallTraceEntry[{functionName}]: activation succeeded via {targetLabel}/{clickLabel}");
                        return;
                    }
                    catch (TimeoutException)
                    {
                        DebugLogger.Log($"CallTraceEntry[{functionName}]: selection timeout after {targetLabel}/{clickLabel}");
                    }
                }
                catch (PlaywrightException ex)
                {
                    DebugLogger.Log($"CallTraceEntry[{functionName}]: PlaywrightException on {targetLabel}/{clickLabel}: {ex.Message}");
                }
            }
        }

        DebugLogger.Log($"CallTraceEntry[{functionName}]: failed to activate via all targets");
        throw new InvalidOperationException("Failed to activate call trace entry via any known target.");
    }

    /// <summary>
    /// Retrieves all argument descriptors rendered for the call.
    /// </summary>
    public async Task<IReadOnlyList<CallTraceArgument>> ArgumentsAsync()
    {
        var args = await ArgumentContainer().Locator(".call-arg").AllAsync();
        return args.Select(locator => new CallTraceArgument(_pane, locator, _contextMenu)).ToList();
    }

    /// <summary>
    /// Retrieves the return value text if available, otherwise null.
    /// </summary>
    public async Task<string?> ReturnValueAsync()
    {
        if (await ReturnLocator().CountAsync() == 0)
        {
            return null;
        }

        var text = await ReturnLocator().First.InnerTextAsync();
        return text?.Trim();
    }
}

File: tsc-ui-tests/page-objects/panes/call-trace/call-trace-entry.ts
import type { Locator } from "@playwright/test";
import { ContextMenu } from "../../components/context-menu";
import { CallTraceArgument } from "./call-trace-argument";
import type { CallTracePane } from "./call-trace-pane";
import { retry } from "../../../lib/retry-helpers";
import { debugLogger } from "../../../lib/debug-logger";

const EXPAND_CHILDREN_OPTIONS = ["Expand Call Children", "Expand Full Callstack"];
const COLLAPSE_CHILDREN_OPTIONS = ["Collapse Call Children", "Expand Full Callstack"];

const ACTIVATE_RETRY_ATTEMPTS = 5;
const ACTIVATE_RETRY_DELAY_MS = 50;
const ACTIVATE_DOUBLE_CLICK_DELAY_MS = 50;

/**
 * Represents a single call entry rendered within the call trace pane.
 *
 * Port of ui-tests/PageObjects/Panes/CallTrace/CallTraceEntry.cs
 */
export class CallTraceEntry {
  private readonly pane: CallTracePane;
  readonly root: Locator;
  private readonly contextMenu: ContextMenu;

  constructor(pane: CallTracePane, root: Locator, contextMenu: ContextMenu) {
    this.pane = pane;
    this.root = root;
    this.contextMenu = contextMenu;
  }

  private offsetLocator(): Locator {
    return this.root.locator("> span").first();
  }

  private childBoxLocator(): Locator {
    return this.root.locator(".call-child-box");
  }

  private callTextLocator(): Locator {
    return this.childBoxLocator().locator(".call-text");
  }

  private toggleLocator(): Locator {
    return this.childBoxLocator().locator(".toggle-call");
  }

  private returnLocator(): Locator {
    return this.childBoxLocator().locator(".return-text");
  }

  private argumentContainer(): Locator {
    return this.childBoxLocator().locator(".call-args");
  }

  async callText(): Promise<string> {
    return (await this.callTextLocator().innerText()).trim();
  }

  /**
   * Extracts the function name from the call text.
   * Handles Ruby-style "Class#method #N" by finding the last " #N" pattern.
   */
  async functionName(): Promise<string> {
    const callText = await this.callText();
    for (let i = callText.length - 1; i >= 2; i--) {
      if (
        callText[i - 1] === "#" &&
        callText[i - 2] === " " &&
        /\d/.test(callText[i])
      ) {
        return callText.substring(0, i - 2).trim();
      }
    }
    return callText;
  }

  async callIdentifier(): Promise<string> {
    const callText = await this.callText();
    for (let i = callText.length - 1; i >= 2; i--) {
      if (
        callText[i - 1] === "#" &&
        callText[i - 2] === " " &&
        /\d/.test(callText[i])
      ) {
        return callText.substring(i).trim();
      }
    }
    return "";
  }

  /**
   * Depth inferred from the leading spacer element's min-width CSS.
   */
  async depth(): Promise<number> {
    const style = (await this.offsetLocator().getAttribute("style")) ?? "";
    const match = style.match(/min-width:\s*([\d.]+)px/);
    if (!match) return 0;
    return Math.round(parseFloat(match[1]) / 8);
  }

  private async hasClass(className: string): Promise<boolean> {
    const classAttr = (await this.root.getAttribute("class")) ?? "";
    return classAttr.split(/\s+/).includes(className);
  }

  async isSelected(): Promise<boolean> {
    return this.hasClass("event-selected");
  }

  async hasToggle(): Promise<boolean> {
    return (
      (await this.toggleLocator()
        .locator(".collapse-call-img, .dot-call-img")
        .count()) > 0
    );
  }

  async hasExpandedChildren(): Promise<boolean> {
    return (await this.toggleLocator().locator(".collapse-call-img").count()) > 0;
  }

  async expectedContextMenu(): Promise<readonly string[]> {
    return (await this.hasExpandedChildren())
      ? COLLAPSE_CHILDREN_OPTIONS
      : EXPAND_CHILDREN_OPTIONS;
  }

  async contextMenuEntries(): Promise<string[]> {
    await this.callTextLocator().click({ button: "right" });
    await this.contextMenu.waitForVisible();
    const entries = await this.contextMenu.getEntries();
    await this.contextMenu.dismiss();
    return entries.map((e) => e.text);
  }

  async expandChildren(): Promise<void> {
    if (!(await this.hasToggle())) return;
    if (!(await this.hasExpandedChildren())) {
      await this.toggleLocator().click();
      await retry(() => this.hasExpandedChildren());
    }
  }

  async collapseChildren(): Promise<void> {
    if (!(await this.hasToggle())) return;
    if (await this.hasExpandedChildren()) {
      await this.toggleLocator().click();
      await retry(async () => !(await this.hasExpandedChildren()));
    }
  }

  /**
   * Clicks the call entry to trigger jump navigation.
   * Uses multiple strategies to work around viewport/rendering issues on Windows.
   */
  async activate(): Promise<void> {
    const functionName = await this.functionName();
    debugLogger.log(`CallTraceEntry[${functionName}]: Begin activate`);

    const callText = this.callTextLocator();

    // Strategy 1: Use page.evaluate to directly click the DOM element
    // This bypasses all Playwright viewport/actionability checks
    try {
      debugLogger.log(`CallTraceEntry[${functionName}]: evaluate click on call-text`);
      await callText.evaluate((el: HTMLElement) => el.click());
      await this.pane.page.waitForTimeout(500);
      if (await this.isSelected()) {
        debugLogger.log(`CallTraceEntry[${functionName}]: activated via evaluate click`);
        return;
      }
    } catch {
      debugLogger.log(`CallTraceEntry[${functionName}]: evaluate click failed`);
    }

    // Strategy 2: dispatchEvent("click")
    try {
      debugLogger.log(`CallTraceEntry[${functionName}]: dispatchEvent click`);
      await callText.dispatchEvent("click");
      await this.pane.page.waitForTimeout(500);
      if (await this.isSelected()) {
        debugLogger.log(`CallTraceEntry[${functionName}]: activated via dispatchEvent`);
        return;
      }
    } catch {
      debugLogger.log(`CallTraceEntry[${functionName}]: dispatchEvent click failed`);
    }

    // Strategy 3: scrollIntoView + force click with short timeout
    try {
      await callText.scrollIntoViewIfNeeded({ timeout: 2_000 });
    } catch {
      // ignore scroll failure
    }
    try {
      debugLogger.log(`CallTraceEntry[${functionName}]: force click`);
      await callText.click({ force: true, timeout: 5_000 });
      await this.pane.page.waitForTimeout(500);
      if (await this.isSelected()) {
        debugLogger.log(`CallTraceEntry[${functionName}]: activated via force click`);
        return;
      }
    } catch {
      debugLogger.log(`CallTraceEntry[${functionName}]: force click failed`);
    }

    // If none of the above set the selection, log a warning but don't throw.
    // The caller may still succeed (e.g., the navigation jump already happened
    // as a side effect of the search result click).
    debugLogger.log(
      `CallTraceEntry[${functionName}]: WARNING: entry not selected after all strategies, proceeding anyway`,
    );
  }

  async arguments(): Promise<CallTraceArgument[]> {
    const args = await this.argumentContainer().locator(".call-arg").all();
    return args.map((l) => new CallTraceArgument(this.pane, l, this.contextMenu));
  }

  async returnValue(): Promise<string | null> {
    if ((await this.returnLocator().count()) === 0) return null;
    return (await this.returnLocator().first().innerText()).trim();
  }
}
